bindings/pyroot_experimental/PyROOT/src/DeclareCallablePyz.cxx

```diff
@@ -0,0 +1,663 @@
+   code << "   // Build Python proxies for C++ objects\n";
+   std::stringstream cleanup; // Register clean-up code
+   bool hasPyBoolIncref = false;
+   for (std::size_t i = 0; i < pytypes.size(); i++) {
+      if (inputTypesStr[i].compare("bool") == 0) { // Bool
+         if (!hasPyBoolIncref) {
+            code << "   Py_INCREF(Py_True);\n"
+                 << "   Py_INCREF(Py_False);\n";
+            cleanup << "   Py_DECREF(Py_True);\n"
+                    << "   Py_DECREF(Py_False);\n";
+            hasPyBoolIncref = true;
+         }
+         code << "   auto pyb_" << i << " = b_" << i << " ? Py_True : Py_False;\n";
+      } else if (pytypes[i] == "O") { // C++ objects
+         code << "   auto pyo_" << i << " = TPython::CPPInstance_FromVoidPtr("
+              << "&o_" << i << ", \"" << inputTypesStr[i] << "\");\n";
+         cleanup << "   Py_DECREF(pyo_" << i << ");\n";
+      }
+   }
+   code << "\n";
+
+   // Call Python callable
+   auto typestr_str = typestr.str();
+   auto vars_str = vars.str();
+   if (vars_str.compare("") != 0) {
+      vars_str = ", " + vars_str;
+   }
+   code << "   // Call Python callable\n"
+        << "   auto pyresult = PyObject_CallFunction(pyfunc, (char*)\"" << typestr_str << "\"" << vars_str << ");\n"
+        << "   if (pyresult == 0) {\n"
+        << "      PyErr_Print();\n"
+        << "      throw std::runtime_error(\"Failed to call Python callable " << name << ".\");\n"
+        << "   }\n\n";
+
+   // Clean-up Python proxies
+   code << "   // Clean-up Python proxies\n"
+        << cleanup.str()
+        << "\n";
+
+   // Convert result to C++ type
+   code << "   // Convert result to C++ type\n";
+   if (returnTypeStr.compare("void") == 0) {
+      code << "   Py_DECREF(pyresult);\n\n"
+           << "   return;\n";
+   } else if (returnTypeStr.compare("bool") == 0) {
+      code << "   if (pyresult == Py_True) {\n"
+           << "      Py_DECREF(pyresult);\n"
+           << "      return true;\n"
+           << "   } else if (pyresult == Py_False) {\n"
+           << "      Py_DECREF(pyresult);\n"
+           << "      return false;\n"
+           << "   } else {\n"
+           << "      PyErr_Print();\n"
+           << "      throw std::runtime_error(\"Failed to convert return value of Python callable to C++ object: Python object is not of type PyBool.\");\n"
+           << "   }\n";
+   } else {
+      auto pytype = typemap.find(returnTypeStr);
+      if (pytype != typemap.end()) {
+         if (pytype->second == "f") {
+            code << "   auto result = PyFloat_AsDouble(pyresult);\n";
+         } else if (pytype->second == "i" || pytype->second == "l") {
+            code << "   auto result = PyLong_AsLong(pyresult);\n";
+         } else if (pytype->second == "I" || pytype->second == "k") {
+            code << "   auto result = PyLong_AsUnsignedLong(pyresult);\n";
+         }
+      } else {
+         code << "   if (!TPython::CPPInstance_Check(pyresult)) {\n"
```

Comment:
If this happens it would be actually a bug in Cppyy :) But does not hurt to keep the check.

Reply:
Yes, that's right :D